src/liballoc_system/lib.rs

```diff
@@ -96,8 +96,10 @@ mod imp {
             libc::realloc(ptr as *mut libc::c_void, size as libc::size_t) as *mut u8
         } else {
             let new_ptr = allocate(size, align);
-            ptr::copy(ptr, new_ptr, cmp::min(size, old_size));
-            deallocate(ptr, old_size, align);
+            if !new_ptr.is_null() {
+                ptr::copy(ptr, new_ptr, cmp::min(size, old_size));
+                deallocate(ptr, old_size, align);
```

Comment:
Oh, never mind me.

The manual for `realloc` says:

> If the area pointed to was moved, a free(ptr) is done.

So it is required for memory to not get moved if move fails.